[BatchNorm num_batches_tracked update]

jjsjann123 opened the pull request:
1. remove in-place operator on num_batches_tracked during forward calls:
     a. in-place operator on an input param puts heavy contraint on data
        dependency, as AliasDb assumes all input to alias each other.
     b. moreover, currently we incrementing on a scalar instead of tensor, which
        in-place operator is not supported in scripting.
2. move computation onto a scalar instead of a tensor, this saves a kernel
launch when we have the layer on GPU.

jjsjann123 committed:
[BatchNorm num_batches_tracked update]

1. remove in-place operator on num_batches_tracked during forward calls:
     a. in-place operator on an input param puts heavy contraint on data
        dependency, as AliasDb assumes all input to alias each other.
     b. moreover, currently we incrementing on a scalar instead of tensor, which
        in-place operator is not supported in scripting.
2. move computation onto a scalar instead of a tensor, this saves a kernel
launch when we have the layer on GPU.

jjsjann123 commented:
cc'ing @ngimel @mruberry @csarofeen

jjsjann123 committed:
fixing a previous typo

mruberry commented:
Cool!

Do you have some before/after performance numbers for this change?

jjsjann123 committed:
passing CI tests

jjsjann123 committed:
fixing lint issue

jjsjann123 commented:
![image](https://user-images.githubusercontent.com/3709243/65468394-c22fba80-de18-11e9-88de-ceff1360a3c0.png)

My monkey benchmark show general speedup comparing the before and after.
Script that I used:
```
import torch
import itertools
import time

ITER=100

def bn_benchmark(N, C, S):
  #bn = torch.nn.BatchNorm2d(C, momentum=None).cuda()
  bn = torch.nn.BatchNorm2d(C).cuda()
  inp = torch.randn(N, C, S, S, device="cuda")
  o = bn(inp)
  torch.cuda.synchronize()
  start = time.time()
  for i in range(0, ITER):
    o = bn(inp)
    torch.cuda.synchronize()
  #torch.cuda.synchronize()
  return time.time() - start

N = [32, 64, 128]
C = [2, 4, 8, 16, 32, 64, 128]
S = [7, 14, 28, 56, 112]

for n, c, s in itertools.product(N, C, S):
    print("input size: {0}, {1}, {2}, {2} time: {3}".format(n, c, s, bn_benchmark(n, c, s)))
```

jjsjann123 committed:
replacing `__getattr__' rewrite with `@property` per ngimel's comment.

ngimel commented on torch/nn/modules/batchnorm.py:
```diff
@@ -38,27 +38,37 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True,
             init.zeros_(self.bias)
 
+    @property
+    def num_batches_tracked(self):
```
can you have both a buffer and a @property with the same name? That's pretty confusing

ngimel commented on torch/nn/modules/batchnorm.py:
```diff
@@ -69,9 +79,9 @@ def forward(self, input):
             # TODO: if statement only here to tell the jit to skip emitting this when it is None
             if self.num_batches_tracked is not None:
-                self.num_batches_tracked += 1
+                self.num_batches_tracked_scalar = self.num_batches_tracked_scalar + 1
```
Not that it matters, but you can also do += here

jjsjann123 replied:
`+=` actually breaks jit.script

JIT support for in-place operator on scalar is limited. `left-hand side of augmented assignment to module parameters/buffers can only be tensor types`

ngimel reviewed (changes requested):
This looks good, but I'm confused by similarly named @property and buffer. If it indeed works, a comment would be nice. Also, are there existing tests testing saving and loading batchnorm? If not, it would make sense to add them, to make sure checkpoints created before this PR are still loadable after. Does this change merit bumping the _version?

jjsjann123 commented:
__getattr__ plays fine with @property (since __getattr__ is only triggered when we access something that doesn't exist in __dict__). But I didn't work out how to support manually over-write num_batched_tracked.

As that goes through the __setattr__ in torch.nn.Module which redirects the write to update OrderedDict _buffer. So I can't hijack it with @property.setter.

Discussed it briefly offline with @ngimel , we decide to bump the _version, where we can have `num_batches_tracked` as a scalar, instead of being a tensor.
We'll work it around inside BN layer so it will be compatible with loading older saved module.
I'll update this PR shortly.

jjsjann123 committed:
[BatchNorm num_batches_tracked update]

New implementation, changed _BatchNorm.num_batches_tracked to an integer instead
of a tensor.

But we kept the exported `state_dict` the same way that we convert it to a
tensor when exporting. This avoids touching many code in our current testing
infrastructure which uses Tensor APIs.

jjsjann123 commented:
ended up not bumping `_version`, as we wrap `num_batches_tracked` inside a tensor in `state_dict`. This plays nicely with nn tests, as it assumes all value inside `state_dict` supports tensor APIs.

jjsjann123 commented:
Also, there are nn tests for BC: https://github.com/pytorch/pytorch/pull/26550/files#diff-d89baec73022f5f511c5beb5ce6498dfR4502-R4517

ssnl reviewed (commented):
Sorry for the late review. Why did we abandon the `@property` approach? 

Another concern is that you can not now set `num_batch_tracked` easily in a BC compatible way. Maybe `attr.setter` and `attr.getter` is fine?

jjsjann123 commented:
`property.setter` won't work without us re-write `__setattr__` and we figure it will be too invasive.

So basically, as long as we maintain `num_batches_tracked` as a buffer/parameter in `_BatchNorm` (https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L626), our property setter be triggered simply because `__setattr__` hijacked the call.

We can get it to work if we change the name for the actual stored buffer/parameter, and keep a `num_batches_tracked` as a property in `_BatchNorm`, that way we can preserve the API to be somewhat BC, but it will still behave differently.
e.g. 
```
n = torch.tensor([0])
bn.num_batches_tracked = n
n++
print(bn.num_batches_tracked)
```
Where in master PyTorch now, it prints out 1, but we can't do that in either implementation proposed in this PR, simply because that we are removing the in-place operation.

I'm open minded on the choice that we make.
IMHO, it will be better and less confusion to not let user pass in a tensor from the beginning.

jjsjann123 commented:
The impact of breaking BC is bigger than what I was expecting. We can probably fix this by going through all the JIT issues that was exposed in CI failures, but I'm non sure about having BN as an exception here, where part of the parameter/buffer is stored as a scalar other than a tensor.

For now we decide to temporarily drop this PR until we got more time to investigate and properly fix it.
I'll open a one line PR to remove the in-place operation on `num_batches_tracked`, so data dependency does not isolate the graph during JIT optimization passes.

Should we close this PR and track our attempt in an issue?

jjsjann123 committed:
proper API to support assignment of tensor as well as scalar to num_batches_tracked

ngimel commented:
Closing per author's request, open an issue please.
